homeassistant/components/sensor/mqtt.py

```diff
@@ -74,15 +82,36 @@ def async_added_to_hass(self):
         @callback
         def message_received(topic, payload, qos):
             """A new MQTT message has been received."""
+            # auto-expire enabled?
+            if self._expire_after > 0:
+                # Reset expiration time and set trigger
+                self._value_expiration_at = (
+                    dt_util.utcnow() +
+                    timedelta(seconds=self._expire_after)
+                    )
+                async_track_point_in_utc_time(self.hass,
```

Comment:
`async_track_point_in_utc_time` returns a function that will unsubscribe the just subscribed listener. So instead of storing a timestamp when to expire, unsubscribe the old listener if it exists and subscribe a new one.

Then in the callback, set the unsubscribe function to `None`